File: DynamicProgramming/HouseRobber/rob.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int findMaxRob(vector<int>& nums, vector<int>& dp, int idx){
        if(idx == 0)return nums[idx];
        else if(idx == 1)return max(nums[idx],nums[idx-1]);
        else if(dp[idx]!=-1)return dp[idx];
        return dp[idx] = max(findMaxRob(nums,dp,idx-1),nums[idx]+findMaxRob(nums,dp,idx-2));
    }
    int rob(vector<int>& nums) {
        int n = nums.size();
        vector<int>dp(n,-1);
        dp[0] = nums[0];
        return findMaxRob(nums,dp,n-1);
    }
};

File: DynamicProgramming/HouseRobber/rob.py
class Solution:  
    def findMaxSum(self, arr):
        # code here
        n = len(arr)
        if n==1: return arr[0]
        arr[1] = max(arr[0],arr[1])
        for i in range(2,n):
            arr[i] = max(arr[i-1],arr[i-2]+arr[i])
        return arr[n-1]